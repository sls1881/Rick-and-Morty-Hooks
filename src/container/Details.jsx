import React, { useState, useEffect} from 'react'
import { useParams } from 'react-router'
import DetailsItem from '../components/app/characters/DetailsItem'
import Spinner from '../components/Spinner'
import { getCharacter } from '../services/ApiUtils'

export default function Details() {
    //state
    const [loading, setLoading] = useState(true)
    const [character, setCharacter] = useState({})
    
    const {id} = useParams();

    //useEffect
    useEffect(() => {
        (async () => {
            const character = await getCharacter(id)
            //setCharacter, setLoading
            setCharacter(character);
            setLoading(false);
        })();
    }, [id]);

    return (
        <div>
            <h1>Rick and Morty Character</h1>
            {loading ? <Spinner/> : <DetailsItem character={character}/>}
        </div>
    )
}
