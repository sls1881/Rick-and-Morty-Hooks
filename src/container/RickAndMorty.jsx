import React, { useState, useEffect } from 'react'
import CharacterList from '../components/app/characters/CharacterList';
import Spinner from '../components/Spinner';
import styles from '../components/app/App.css';
import { getCharacters } from '../services/ApiUtils';

export default function RickAndMorty() {
//useState
const [characters, setCharacters] = useState([]);
const [loading, setLoading] = useState(true);
const [page, setPage] = useState(1);

//useEffect and fetch
useEffect(() => {
    (async () => {
        const characters = await getCharacters(page)
        //setCharacter, setLoading, setPage
        setCharacters(characters);
        setLoading(false);
    })();
}, [page]);

    return (
        <div>
            <div className={styles.characterContainer}>
            {loading ? <Spinner/> : <CharacterList characters={characters}/>}
            </div>
            <div className={styles.buttons}>
            <button onClick={() => setPage((prevPage) => prevPage - 1)}>Previous</button>
            <span>{page}</span>
            <button aria-label='pages' onClick={() => setPage((nextPage) => nextPage + 1)}>Next</button>
            </div>
        </div>
    )
}
