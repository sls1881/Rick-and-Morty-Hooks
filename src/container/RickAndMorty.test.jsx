import React from 'react'
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import RickAndMorty from './RickAndMorty';
import mockData from './chracters.json'
import { MemoryRouter, Route } from 'react-router';
import Details from './Details';

const server = setupServer(rest.get('https://rickandmortyapi.com/api/character', (req, res, ctx) => {
    return res(ctx.json(mockData))
}))

describe('Rick and Morty API', () => {
    beforeAll(() => server.listen());
    afterAll(() => server.close());

    it('Render a list of characters to the page', async () => {
render(<MemoryRouter><RickAndMorty/></MemoryRouter>)
        screen.getByAltText('spinner')

        const ul = await screen.findByRole('list', {name: 'characters'})
        expect(ul).toMatchSnapshot();

        const button = await screen.findByRole('button', { name: 'pages'})
        userEvent.click(button);

        waitFor(() => {
            screen.getByAltText('Rick Sanchez')
            screen.getByText('Human')
        })
    })

    it('Render a single character', async () => {
        render(<MemoryRouter initialEntries={['/3']}>
        <Route path='/:id'>
            <Details />
        </ Route>
    </MemoryRouter>) 

        screen.getByAltText('spinner');
        const ul = await screen.findByRole('list', {name: 'character'})
        expect(ul).toMatchSnapshot();

        return waitFor(() => {
            screen.getByAltText('Summer Smith')
            screen.getByText('Alive');
        })
    })
})