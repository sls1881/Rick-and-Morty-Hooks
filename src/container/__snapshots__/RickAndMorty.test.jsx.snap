// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rick and Morty API Render a list of characters to the page 1`] = `
<ul
  aria-label="characters"
>
  <li>
    <a
      class="link"
      href="/1"
    >
      <div>
        <h2>
          Rick Sanchez
        </h2>
        <img
          alt="Rick Sanchez"
          src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/2"
    >
      <div>
        <h2>
          Morty Smith
        </h2>
        <img
          alt="Morty Smith"
          src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/3"
    >
      <div>
        <h2>
          Summer Smith
        </h2>
        <img
          alt="Summer Smith"
          src="https://rickandmortyapi.com/api/character/avatar/3.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/4"
    >
      <div>
        <h2>
          Beth Smith
        </h2>
        <img
          alt="Beth Smith"
          src="https://rickandmortyapi.com/api/character/avatar/4.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/5"
    >
      <div>
        <h2>
          Jerry Smith
        </h2>
        <img
          alt="Jerry Smith"
          src="https://rickandmortyapi.com/api/character/avatar/5.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/6"
    >
      <div>
        <h2>
          Abadango Cluster Princess
        </h2>
        <img
          alt="Abadango Cluster Princess"
          src="https://rickandmortyapi.com/api/character/avatar/6.jpeg"
        />
        <p>
          Alien
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/7"
    >
      <div>
        <h2>
          Abradolf Lincler
        </h2>
        <img
          alt="Abradolf Lincler"
          src="https://rickandmortyapi.com/api/character/avatar/7.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/8"
    >
      <div>
        <h2>
          Adjudicator Rick
        </h2>
        <img
          alt="Adjudicator Rick"
          src="https://rickandmortyapi.com/api/character/avatar/8.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/9"
    >
      <div>
        <h2>
          Agency Director
        </h2>
        <img
          alt="Agency Director"
          src="https://rickandmortyapi.com/api/character/avatar/9.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/10"
    >
      <div>
        <h2>
          Alan Rails
        </h2>
        <img
          alt="Alan Rails"
          src="https://rickandmortyapi.com/api/character/avatar/10.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/11"
    >
      <div>
        <h2>
          Albert Einstein
        </h2>
        <img
          alt="Albert Einstein"
          src="https://rickandmortyapi.com/api/character/avatar/11.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/12"
    >
      <div>
        <h2>
          Alexander
        </h2>
        <img
          alt="Alexander"
          src="https://rickandmortyapi.com/api/character/avatar/12.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/13"
    >
      <div>
        <h2>
          Alien Googah
        </h2>
        <img
          alt="Alien Googah"
          src="https://rickandmortyapi.com/api/character/avatar/13.jpeg"
        />
        <p>
          Alien
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/14"
    >
      <div>
        <h2>
          Alien Morty
        </h2>
        <img
          alt="Alien Morty"
          src="https://rickandmortyapi.com/api/character/avatar/14.jpeg"
        />
        <p>
          Alien
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/15"
    >
      <div>
        <h2>
          Alien Rick
        </h2>
        <img
          alt="Alien Rick"
          src="https://rickandmortyapi.com/api/character/avatar/15.jpeg"
        />
        <p>
          Alien
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/16"
    >
      <div>
        <h2>
          Amish Cyborg
        </h2>
        <img
          alt="Amish Cyborg"
          src="https://rickandmortyapi.com/api/character/avatar/16.jpeg"
        />
        <p>
          Alien
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/17"
    >
      <div>
        <h2>
          Annie
        </h2>
        <img
          alt="Annie"
          src="https://rickandmortyapi.com/api/character/avatar/17.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/18"
    >
      <div>
        <h2>
          Antenna Morty
        </h2>
        <img
          alt="Antenna Morty"
          src="https://rickandmortyapi.com/api/character/avatar/18.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/19"
    >
      <div>
        <h2>
          Antenna Rick
        </h2>
        <img
          alt="Antenna Rick"
          src="https://rickandmortyapi.com/api/character/avatar/19.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
  <li>
    <a
      class="link"
      href="/20"
    >
      <div>
        <h2>
          Ants in my Eyes Johnson
        </h2>
        <img
          alt="Ants in my Eyes Johnson"
          src="https://rickandmortyapi.com/api/character/avatar/20.jpeg"
        />
        <p>
          Human
        </p>
      </div>
    </a>
    <hr />
  </li>
</ul>
`;

exports[`Rick and Morty API Render a single character 1`] = `
<ul
  aria-label="character"
>
  <li>
    <h2>
      Summer Smith
    </h2>
    <p>
      Human
    </p>
    <p>
      Alive
    </p>
    <img
      alt="Summer Smith"
      src="https://rickandmortyapi.com/api/character/avatar/3.jpeg"
    />
    <a
      href="https://rickandmortyapi.com/api/character/3"
    >
      <p>
        Character Data
      </p>
    </a>
  </li>
</ul>
`;
